isabelle/Core/Semantics.thy

```diff
@@ -41,22 +41,11 @@ evalValue_SubValue: "evalValue env state (SubValue lhs rhs) =
     evalValue env state lhs - evalValue env state rhs" |
 evalValue_MulValue: "evalValue env state (MulValue lhs rhs) =
     evalValue env state lhs * evalValue env state rhs" |
-evalValue_DivValue: "evalValue env state (DivValue lhs rhs) = 
-   (let n = evalValue env state lhs in 
-    if n = 0 then 0 
-    else let d = evalValue env state rhs in 
-      if d = 0 then 0 else 
-        let (q, r) = n quotRem d in 
-        let ar = abs r * 2 in
-        let ad = abs d in 
-        if ar < ad then q
-        else if ar > ad then q + signum n * signum d
-        else let qIsEven = q rem 2 = 0 in 
-          if qIsEven then q else q + signum n * signum d)" |
-evalValue_Scale: "evalValue env state (Scale n d rhs) = 
-    (let nn = evalValue env state rhs * n in
-     let (q, r) = nn quotRem d in
-     if abs r * 2 < abs d then q else q + signum nn * signum d)" |
+evalValue_DivValue: "evalValue env state (DivValue lhs rhs) =
+   (let n = evalValue env state lhs;
+        d = evalValue env state rhs
+    in if d = 0 then 0
```

Comment:
Division by zero is defined as zero - is that necessary? (It has also been the case for financial rounding).

Reply:
Yes, the `quot` function will throw an exception in that situation and crash, and we want semantics to be complete and fault tolerant to avoid locking funds. Ideally, we would issue a warning too but that would complicate the implementation of evalValue, so we decided in the past to skip that for simplicity